# Assignment 1 vs. Assignment 2
**Comparing `output.txt` with `output_rdd.txt`**

We changed the regex between Assignment 1 and Assignment 2 to better match the delimiter specification in the assignment.
In Assignment 1 we used:
```Python
re.compile(r"[^a-zA-Z<>^|]+")
```
This regex splits on every sequence of characters that is not an uppercase/lowercase letter and not one of `<`, `>`, `^`, or `|`. In other words, it defines what is allowed inside a token. Everything outside that allowed set becomes a delimiter. This is stricter because tokens can only contain letters plus those few special characters.

In Assignment 2 we used:
```Python
re.compile(r'[\s\d()\[\]{}.!?,;:+=\-_"\'`~#@&*%€$§\\/]+')
```

This regex directly lists the delimiter characters from the assignment: whitespace, digits, and punctuation such as brackets, dots, commas, quotes, currency symbols, slash, etc. This means it defines what should be used to split the text, instead of defining which characters are allowed in a token.

```
Assignment 1 regex:
    keeps only letters plus < > ^ |
    splits on everything else

Assignment 2 regex:
    splits only on the assignment’s specified delimiters
    keeps anything that is not one of those delimiters
```

In [30]:
ass1_output_filename = "output.txt"
ass2_part1_output_filename = "output_rdd.txt"
# read files
with open(ass1_output_filename, "r") as f:
    ass1_output = f.read()
    ass1_output_lines = ass1_output.splitlines()
    ass1_output_categories = []
    ass1_output_data = {
        "terms":[ w for w in ass1_output_lines.pop().split(" ")]
    }
    for line in ass1_output_lines:
        category, chi2values = line.split(" ", 1)
        ass1_output_categories.append(category)
        for term_count in chi2values.split(" "):
            term, chi2values = term_count.split(":")
            ass1_output_data.setdefault(category, {})[term] = float(chi2values)
            
with open(ass2_part1_output_filename, "r") as f:
    ass2_part1_output = f.read()
    ass2_part1_output_lines = ass2_part1_output.splitlines()
    ass2_part1_output_categories = []
    ass2_part1_output_data = {
        "terms":[ w for w in ass2_part1_output_lines.pop().split(" ")]
    }
    for line in ass2_part1_output_lines:
        category, chi2values = line.split(" ", 1)
        ass2_part1_output_categories.append(category)
        for term_count in chi2values.split(" "):
            term, chi2values = term_count.split(":")
            ass2_part1_output_data.setdefault(category, {})[term] = float(chi2values)

In [31]:
# Compare the two categories, the terms and their order and their chi2 values
# First check if categories are the same
if set(ass1_output_categories) != set(ass2_part1_output_categories):
    print("Categories differ between Assignment 1 and Part 1")
else:
    print("Categories are the same between Assignment 1 and Part 1")
    # Check if terms are the same for each category
    for category in ass1_output_categories:
        ass1_terms = set(ass1_output_data[category].keys())
        ass2_terms = set(ass2_part1_output_data[category].keys())
        if ass1_terms != ass2_terms:
            print(f"Terms differ for category {category}")
        else:
            print(f"Terms are the same for category {category}")

Categories are the same between Assignment 1 and Part 1
Terms are the same for category Apps_for_Android
Terms are the same for category Automotive
Terms are the same for category Baby
Terms are the same for category Beauty
Terms are the same for category Book
Terms are the same for category CDs_and_Vinyl
Terms are the same for category Cell_Phones_and_Accessorie
Terms are the same for category Clothing_Shoes_and_Jewelry
Terms are the same for category Digital_Music
Terms are the same for category Electronic
Terms are the same for category Grocery_and_Gourmet_Food
Terms are the same for category Health_and_Personal_Care
Terms are the same for category Home_and_Kitche
Terms are the same for category Kindle_Store
Terms are the same for category Movies_and_TV
Terms are the same for category Musical_Instrument
Terms are the same for category Office_Product
Terms are the same for category Patio_Lawn_and_Garde
Terms are the same for category Pet_Supplie
Terms are the same for category Sports

Only for Category **Tools_and_Home_Improvement** the terms do not match.

In [32]:
# Show the difference in the terms for category Tools_and_Home_Improvement
category = "Tools_and_Home_Improvement"
ass1_terms = set(ass1_output_data[category].keys())
ass2_terms = set(ass2_part1_output_data[category].keys())
ass1_wrong_terms = ass1_terms - ass2_terms
ass2_wrong_terms = ass2_terms - ass1_terms
print(f"Terms in Assignment 1 but not in Part 1 for category {category}: {ass1_wrong_terms}")
print(f"Terms in Part 1 but not in Assignment 1 for category {category}: {ass2_wrong_terms}")
# Show the difference in the chi2 values for the wrong terms
for term in ass1_wrong_terms:
    print(f"Chi2 value for term {term} in Assignment 1: {ass1_output_data[category][term]}")
for term in ass2_wrong_terms:
    print(f"Chi2 value for term {term} in Part 1: {ass2_part1_output_data[category][term]}")

Terms in Assignment 1 but not in Part 1 for category Tools_and_Home_Improvement: {'kidde'}
Terms in Part 1 but not in Assignment 1 for category Tools_and_Home_Improvement: {'bidet'}
Chi2 value for term kidde in Assignment 1: 159.72357729489363
Chi2 value for term bidet in Part 1: 159.72357729489363


Comparing the two results, one can find that both terms for this specific category have the same ChiSquared value, which is why it might happen, that the order of them appearing in the list of words might change, and one word is cut off.

In [33]:
# Now compare all the chi2 values for the same terms and categories
for category in ass1_output_categories:
    for term in ass1_output_data[category].keys():
        if term in ass1_wrong_terms or term in ass2_wrong_terms:
            continue
        chi2_ass1 = ass1_output_data[category][term]
        chi2_ass2 = ass2_part1_output_data[category].get(term, None)
        if chi2_ass2 is None:
            print(f"Term {term} in category {category} is missing in Part 1")
        elif chi2_ass1 != chi2_ass2:
            print(f"Chi2 value for term {term} in category {category} differs: Assignment 1 has {chi2_ass1}, Part 1 has {chi2_ass2}")
            
for category in ass2_part1_output_categories:
    for term in ass2_part1_output_data[category].keys():
        if term in ass1_wrong_terms or term in ass2_wrong_terms:
            continue
        chi2_ass2 = ass2_part1_output_data[category][term]
        chi2_ass1 = ass1_output_data[category].get(term, None)
        if chi2_ass1 is None:
            print(f"Term {term} in category {category} is missing in Assignment 1")
        elif chi2_ass1 != chi2_ass2:
            print(f"Chi2 value for term {term} in category {category} differs: Assignment 1 has {chi2_ass1}, Part 1 has {chi2_ass2}")

**All ChiSquared values for the overlapping terms per category match!**

In [34]:
# Compare the overall terms list
ass1_overall_terms = set(ass1_output_data["terms"])
ass2_overall_terms = set(ass2_part1_output_data["terms"])
ass1_wrong_overall_terms = ass1_overall_terms - ass2_overall_terms
ass2_wrong_overall_terms = ass2_overall_terms - ass1_overall_terms
print(f"Overall terms in Assignment 1 but not in Part 1: {ass1_wrong_overall_terms}")
print(f"Overall terms in Part 1 but not in Assignment 1: {ass2_wrong_overall_terms}")

Overall terms in Assignment 1 but not in Part 1: {'kidde'}
Overall terms in Part 1 but not in Assignment 1: {'bidet'}


These are the same two terms we have a missmatch in the **Tools_and_Home_Improvement** category, which explains the difference!

# Assignment 1 vs. Part 2
**Comparing `output_rdd.txt` with `output_ds.txt`**

In [35]:
ass1_output_data = ass2_part1_output_data["terms"]
ass2_part2_filename = "output_ds.txt"
with open(ass2_part2_filename, "r") as f:
    ass2_part2_output = f.read()
    ass2_part2_output_terms = ass2_part2_output.splitlines()[0].split(" ")
    

In [36]:
#compare the terms in Part 1 and Part 2 and their order
ass1_terms_set = set(ass1_output_data)
ass2_part2_terms_set = set(ass2_part2_output_terms)
print(f"Maximal possible term count for Part 1: {len(ass1_output_categories) * 75}")
print(f"Count of terms in Assignment 1: {len(ass1_terms_set)}")
print(f"Count of terms in Part 2: {len(ass2_part2_terms_set)}")
ass1_wrong_terms = ass1_terms_set - ass2_part2_terms_set
ass2_part2_wrong_terms = ass2_part2_terms_set - ass1_terms_set
print(f"Count of terms in Assignment 1 but not in Part 2: {len(ass1_wrong_terms)}")
print(f"Count of terms in Part 2 but not in Assignment 1: {len(ass2_part2_wrong_terms)}")

# How many terms are overlapping between Part 1 and Part 2
ass1_and_ass2_overlap = ass1_terms_set.intersection(ass2_part2_terms_set)
print(f"Count of terms in both Assignment 1 and Part 2: {len(ass1_and_ass2_overlap)}")

Maximal possible term count for Part 1: 1650
Count of terms in Assignment 1: 1464
Count of terms in Part 2: 2000
Count of terms in Assignment 1 but not in Part 2: 713
Count of terms in Part 2 but not in Assignment 1: 1249
Count of terms in both Assignment 1 and Part 2: 751


**Explanation for the difference:**
The overlap between Part 1 and Part 2 is only partial, which is expected. Part 1 computes chi-square scores manually for each (category, word) pair and keeps the top 75 terms per category, then merges these category-specific lists. Therefore, it favors terms that are highly characteristic for one category, even if they are rare overall.

Part 2 uses Spark ML’s pipeline: CountVectorizer builds a global vocabulary, IDF weights terms corpus-wide, and ChiSqSelector selects the top 2000 features overall with respect to the multiclass category label. This favors globally useful predictive terms rather than the strongest terms per individual category.

Another difference is that Part 1 uses unique word presence per review, while Part 2 uses term-frequency vectors weighted by IDF. Therefore, repeated words and global document frequency influence Part 2 but not Part 1 in the same way.

This explains the observed result: Part 1 has 1464 merged terms, Part 2 has 2000 selected terms, and 751 terms overlap. The non-overlapping terms come from the different selection strategy, feature representation, and ranking objective.